### ISSUE TITLE
Can i create two instance and open single db at read only mode.

### ISSUE THREAD

**Issue by myloft** [cockroachdb/pebble#1263]
I have program A B C。A use to crate db and load data. B and C form as HA. But i find B and C can not open single DB at read only mode. Can you give me some suggestion, or why design like this.

**Comment by petermattis** [comment]
Pebble read-only mode currently grabs an exclusive lock on the DB. We could relax that restriction and let read-only mode grab a shared lock. There is a bit of plumbing of OS specific code in order to achieve this, though.

Stepping back from your immediate question, what benefit do you expect to see from having two processes open the DB in read-only mode. Two processes on the same machine doesn't given any real high-availability. It isn't clear to me why you'd want a separate process to create the DB from the one that reads it. On the surface that seems like more moving parts for no benefit.

**Comment by syncplify** [comment]
On the other hand (and not related to high availability) there are a number of scenarios where it would be greatly beneficial to have 1 process open the DB in read-write mode, and a second (or multiple) processes open the same DB in read-only mode. Single writer, multiple readers.
Though, this is typically a scenario where one would use a DB server, I can see scenarios where this type of access to the key-value store (1-writer-N-readers without an actual DB server) would be greatly beneficial.

**Comment by petermattis** [comment]
@syncplify Can you elaborate on the scenarios where you see a benefit to a single writer process and multiple readers? The one scenario that comes to mind for CRDB is for using CLI tools on a running CRDB node. To date, that hasn't motivated us to do anything in this area.

**Comment by syncplify** [comment]
@petermattis you are correct, there are not many scenarios; in our case we have one main process that's always-on and is the only one that has read-write access to the DB, and several other processes that we call "pop-up processes" that can run independently at unpredictable times, and we know they will always only and exclusively read from (never write to) the DB. Avoiding the overhead of a DB server (or of having the main always-on process act as a pseudo-server) would simplify development by a lot. But on the other hand, we do realize that what we described here is just one of the flavors of a client-server paradigm - an uncommon one, if you will - but still more suited to a true client-server architecture. Therefore, we went ahead and developed a small and lightweight set of functions that allow the always-on process to run the queries for the pop-up processes, de facto acting as a pseudo-server, as you suggested in your previous message. Thank you for the useful exchange of ideas.
